cmd/tools/releaser/releaser.go

```diff
@@ -0,0 +1,238 @@
+				Name:    "set-version",
+				Aliases: []string{"s"},
+				Usage:   "Specific version to set for all modules (overrides --type)",
+			},
+			&cli.BoolFlag{
+				Name:    "prerelease",
+				Aliases: []string{"p"},
+				Usage:   "Create prerelease versions (adds -prereleaseNN suffix). Can be used alone to increment prerelease number only",
+			},
+			&cli.BoolFlag{
+				Name:    "dry-run",
+				Aliases: []string{"d"},
+				Usage:   "Show what would be done without making changes",
+				Value:   true,
+			},
+		},
+		Action: runRelease,
+		Commands: []*cli.Command{
+			{
+				Name:  "examples",
+				Usage: "Show usage examples",
+				Action: func(c *cli.Context) error {
+					showExamples()
+					return nil
+				},
+			},
+		},
+		CustomAppHelpTemplate: `NAME:
+   {{.Name}} - {{.Usage}}
+
+USAGE:
+   {{.HelpName}} [global options]
+
+VERSION:
+   {{.Version}}
+
+GLOBAL OPTIONS:
+   {{range .VisibleFlags}}{{.}}
+   {{end}}
+
+EXAMPLES:
+   {{.HelpName}} --type patch                        # Bump patch version for all modules
+   {{.HelpName}} --type minor --prerelease           # Bump minor version with prerelease suffix
+   {{.HelpName}} --set-version v1.2.3               # Set v1.2.3 for all modules
+   {{.HelpName}} --set-version v1.2.3 --prerelease  # Set v1.2.3-prerelease01 for all modules
+   {{.HelpName}} --prerelease                       # Just increment prerelease number (no version bump)
+   {{.HelpName}} --type major --dry-run              # Show what would happen with major bump
+
+PRERELEASE WORKFLOWS:
+   # Start prerelease cycle for v1.2.4
+   {{.HelpName}} --type patch --prerelease           # Creates v1.2.4-prerelease01
+   
+   # Continue prerelease iterations (no version bump)
+   {{.HelpName}} --prerelease                       # Creates v1.2.4-prerelease02
+   {{.HelpName}} --prerelease                       # Creates v1.2.4-prerelease03
+   
+   # Final release
+   {{.HelpName}} --set-version v1.2.4               # Creates v1.2.4 (removes prerelease suffix)
+
+PRERELEASE FORMAT:
+   Prereleases use STRICT 2-digit numbering (01-99) for proper alphabetical sorting:
+   v1.2.3-prerelease01, v1.2.3-prerelease02, ..., v1.2.3-prerelease99
+   
+   LEGACY SUPPORT: Historical 1-digit prerelease tags are ignored unless they are
+   the latest prerelease for a specific version.
+
+SAFETY FEATURES:
+   - Prevents creating versions that already exist for any module
+   - Verifies builds and runs tests before creating tags
+   - Requires clean git working directory
+   - Enforces releases only from master branch
+
+COMMANDS:
+   {{range .Commands}}{{if not .HideHelp}}   {{join .Names ", "}}{{ "\t"}}{{.Usage}}{{ "\n" }}{{end}}{{end}}
+`,
+	}
+
+	if err := cliApp.Run(os.Args); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func runRelease(c *cli.Context) error {
+	// Parse and validate CLI arguments first
+	cfg := &release.Config{
+		ExcludedDirs:   []string{"cmd", "internal/tools", "idls"},
```

Comment:
I think we might want to tag cmd, as that's how any `go run`-able use will get the code.  not tagging it would allow it to drift.

I think?  I'm thinking of someone who has a go.mod like

```
require (
  github.com/uber/cadence v1.2.3-prerelease01
  github.com/uber/cadence/cmd/server v1.2.3-prerelease01
  github.com/uber/cadence/common/archiver/gcloud v1.2.3-prerelease01
  github.com/uber/cadence/service/sharddistributor/leader/leaderstore/etcd v1.2.3-prerelease01
)
```
which would let them `go run github.com/uber/cadence/cmd/server`, and easily see if versions are in sync.  without that tag I think they'd have to fall back to SHAs for cmd/server.

Reply:
I'm unsure if anyone can use this setup. I'd say people either use docker or fork/modify cmd/server - it is not very friendly package, so unsure if it worth doing.
Though I don't have an opinion, I can add cmd/server to the list of modules.

Reply:
outside temporary dev purposes, I kinda expect that's true, yea.  not sure how to verify that though.

Reply:
Well :) There was no module before, so people either were using master/or not using it.
If we introduce tags, they will use them; otherwise, they will still use master.
Anyway, I've switched to go.work parsing, and now cmd/server is a part of a release.